service/controller/resource/cleanupenis/delete.go

```diff
@@ -2,8 +2,101 @@ package cleanupenis
 
 import (
 	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/v2/pkg/controller/context/finalizerskeptcontext"
+
+	"github.com/giantswarm/aws-operator/service/controller/controllercontext"
 )
 
 func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	cc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var values []*string
+	for _, s := range cc.Status.TenantCluster.TCCP.Subnets {
+		values = append(values, s.SubnetId)
+	}
+
+	var enis []*ec2.NetworkInterface
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding network interfaces")
+
+		i := &ec2.DescribeNetworkInterfacesInput{
+			Filters: []*ec2.Filter{
+				{
+					Name:   aws.String("subnet-id"),
+					Values: values,
+				},
+			},
+		}
+
+		o, err := cc.Client.TenantCluster.AWS.EC2.DescribeNetworkInterfaces(i)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		enis = o.NetworkInterfaces
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d network interfaces", len(enis)))
+	}
+
+	// We want to cleanup network interfaces. We need to check which ENIs are
+	// attached and which are detached. When a network interface is still
+	// attached, its status is in-use. When it is already detached, its status
+	// is available. See e.g. the CLI docs below.
+	//
+	//     https://docs.aws.amazon.com/cli/latest/reference/ec2/wait/network-interface-available.html
+	//
+	var attached []*ec2.NetworkInterface
+	var detached []*ec2.NetworkInterface
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "finding attached and detached network interfaces")
+
+		for _, eni := range enis {
+			switch *eni.Status {
+			case ec2.NetworkInterfaceStatusInUse:
+				attached = append(attached, eni)
+			case ec2.NetworkInterfaceStatusAvailable:
+				detached = append(detached, eni)
+			default:
+				return microerror.Maskf(invalidENIStatusError, *eni.Status)
```

Comment:
I would probabbly just skip here rather than throw error as the status might in transiting state  during the deletion quite often

Reply:
You are right. There are transitional statuses. I will drop the error there. Good call.

Reply:
Sorted.